src/Domain/DomainHelpers.hpp

```diff
@@ -0,0 +1,203 @@
+  static Direction<VolumeDim> mapped(
+      size_t block_id1, size_t block_id2,
+      const Direction<VolumeDim>& dir_in_block1,
+      const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+          corners_of_all_blocks);
+
+  /// Given ids of two blocks, returns the Directions that correspond
+  /// to each block.
+  static std::pair<std::array<Direction<VolumeDim>, VolumeDim>,
+                   std::array<Direction<VolumeDim>, VolumeDim>>
+  obtain_correspondence_between_blocks(
+      size_t block_id1, size_t block_id2,
+      const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+          corners_of_all_blocks);
+
+  /// Given Directions in block1 and their corresponding Directions
+  /// in block2, this function first repackages them into arrays and
+  /// then makes a std::pair of the arrays.
+  static std::pair<std::array<Direction<VolumeDim>, VolumeDim>,
+                   std::array<Direction<VolumeDim>, VolumeDim>>
+  create_correspondence_between_blocks(
+      const Direction<VolumeDim>& direction_to_neighbor_in_self_frame,
+      const Direction<VolumeDim>& direction_to_self_in_neighbor_frame,
+      const std::vector<Direction<VolumeDim>>&
+          alignment_of_shared_face_in_self_frame,
+      const std::vector<Direction<VolumeDim>>&
+          alignment_of_shared_face_in_neighbor_frame);
+
+  /// When setting up periodic-like boundary conditions, the user specifies
+  /// the corners of the two faces to be identified.
+  /// \requires that the two faces correspond to the same block.
+  /// \returns the corresponding id of the block whose faces are to
+  /// be identified.
+  static size_t find_block_id(
+      const std::vector<size_t>& global_corners_face1,
+      const std::vector<size_t>& global_corners_face2,
+      const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+          corners_of_all_blocks);
+
+  /// Sets up the BlockNeighbors using the corner numbering scheme
+  /// to deduce the correct neighbors and orientations. Does not set
+  /// up periodic-like boundary conditions.
+  static void set_internal_boundaries(
+      const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+          corners_of_all_blocks,
+      std::vector<
+          std::unordered_map<Direction<VolumeDim>, BlockNeighbor<VolumeDim>>>&
+          neighbors_of_all_blocks);
+
+  /// Sets up additional BlockNeighbors corresponding to any
+  /// periodic-like boundary condtions provided by the user.
+  static void set_periodic_boundaries(
+      const std::vector<std::vector<size_t>>& identifications,
+      const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+          corners_of_all_blocks,
+      std::vector<
+          std::unordered_map<Direction<VolumeDim>, BlockNeighbor<VolumeDim>>>&
+          neighbors_of_all_blocks);
+};
+
+// ================================================================
+// Implementation details
+// ================================================================
```

Comment:
Move definitions to cpp file, getting rid of the struct. You then will need explicit instantiations at the bottom of the cpp file of `set_external_boundaries` and `set_periodic_boundaries`. The cpp will look like:

```cpp
#include...

namespace {
// implementation function templates (no need to explicitly instantiate)
}  // namespace

template <size_t VolumeDim>
... set_external_boundaries(...) {
...
}

// set_periodic_boundaries

// Explicit instantiations
template ... set_external_boundaries<1>(...);...
```

In hpp have:
```cpp
#include...

/*! Doxygen */
template <size_t VolumeDim>
void set_internal_boundaries(...);

// same for set_periodic_boundaries
```

move doxygen comments for implementation details into cpp file as regular comments